otter/log/cloudfeeds.py

```diff
@@ -0,0 +1,158 @@
+    """
+    cf_event = {}
+    error = False
+
+    # map keys in event to CF keys
+    for log_key, cf_key in log_cf_mapping.iteritems():
+        if log_key in event:
+            cf_event[cf_key] = event[log_key]
+
+    if event.get('isError', False):
+        error = True
+        if ('traceback' in cf_event['message'] or
+           'exception' in cf_event['message']):
+            raise UnsuitableMessage(cf_event['message'])
+
+    return cf_event, error
+
+
+request_format = {
+    "entry": {
+        "@type": "http://www.w3.org/2005/Atom",
+        "title": "autoscale",
+        "content": {
+            "event": {
+                "@type": "http://docs.rackspace.com/core/event",
+                "id": "",
+                "version": "2",
+                "eventTime": "",
+                "type": "INFO",
+                "region": "",
+                "product": {
+                    "@type": "http://docs.rackspace.com/event/autoscale",
+                    "serviceCode": "Autoscale",
+                    "version": "1",
+                    "message": ""
+                }
+            }
+        }
+    }
+}
+
+
+def add_event(event, error, timestamp, region, log, _uuid=uuid.uuid4):
+    """
+    Add event to cloud feeds
+    """
+    request = deepcopy(request_format)
+    if error:
+        request['entry']['content']['event']['type'] = 'ERROR'
+    request['entry']['content']['event']['region'] = region
+    request['entry']['content']['event']['eventTime'] = timestamp
+    request['entry']['content']['event']['product'].update(event)
+    request['entry']['content']['event']['id'] = str(_uuid())
+    return retry_effect(
+        service_request(
+            ServiceType.CLOUD_FEEDS, 'POST',
+            append_segments('autoscale', 'events'), data=request,
+            log=log, success_pred=has_code(201)),
+        retry_times(5), exponential_backoff_interval(2))
+
+
+def add_event_to_cloud_feed(event, timestamp, log=otter_log,
+                            get_disp=get_full_dispatcher):
+    """
+    Add event to cloud feed by sanityzing it first
+
+    :param dict event: Event dict as it would be passed to observer
+    :param str timestamp: ISO8601 formatted timestamp of the event
+    """
+    cf = _cloud_feeds
+    # Take out cloud_feed to avoid coming back here in infinite recursion
+    event.pop('cloud_feed', None)
+    log = log.bind('otter.cloud_feed', **event)
+    try:
+        event, error = sanitize_event(event)
+    except UnsuitableMessage as me:
+        log.err(me, ('Tried to add unsuitable message in cloud feeds: '
+                     '{unsuitable_message}'),
+                unsuitable_message=me.unsuitable_message)
+    else:
+        eff = Effect(TenantScope(cf.tenant_id,
+                                 add_event(event, error, timestamp,
+                                           cf.region, log)))
+        # TODO: Log error if event was not added
+        perform(
```

Comment:
Yeah, everything before the perform() is still close to pure, except:

- `event.pop()`, which can be replaced with `event = dicttoolz.dissoc(event, 'cloud_feed')`
- the `log.err` call, which is not a terrible thing to leave in an otherwise pure function, though you may want to leave it outside at your preference.

So I think add_event is fine, but you could create another function that factors some stuff out of `add_event_to_cloud_feed`. Mainly the TenantScope wrapping will be easier to test if it's in a pure function.